src/ORM/Association/BelongsToMany.php

```diff
@@ -343,30 +347,61 @@ public function attachTo(Query $query, array $options = [])
 
         // Attach the junction table as well we need it to populate _joinData.
         $assoc = $this->_targetTable->association($junction->alias());
-        $query->removeJoin($assoc->name());
-        $options = array_intersect_key($options, ['joinType' => 1, 'fields' => 1]);
-        $options += [
+        $newOptions = array_intersect_key($options, ['joinType' => 1, 'fields' => 1]);
+        $newOptions += [
             'conditions' => $cond,
             'includeFields' => $includeFields,
-            'foreignKey' => $this->targetForeignKey(),
+            'foreignKey' => false,
         ];
-        $assoc->attachTo($query, $options);
+        $assoc->attachTo($query, $newOptions);
         $query->eagerLoader()->addToJoinsMap($junction->alias(), $assoc, true);
+
+        parent::attachTo($query, $options);
+
+        $foreignKey = $this->targetForeignKey();
+        $thisJoin = $query->clause('join')[$this->name()];
+        $thisJoin['conditions']->add($assoc->_joinCondition(['foreignKey' => $foreignKey]));
     }
 
     /**
      * {@inheritDoc}
      */
     protected function _appendNotMatching($query, $options)
     {
-        $target = $this->junction();
-        if (!empty($options['negateMatch'])) {
-            $primaryKey = $query->aliasFields((array)$target->primaryKey(), $target->alias());
-            $query->andWhere(function ($exp) use ($primaryKey) {
-                array_map([$exp, 'isNull'], $primaryKey);
-                return $exp;
-            });
+        if (empty($options['negateMatch'])) {
+            return;
         }
+        $options += ['conditions' => []];
+        $junction = $this->junction();
+        $belongsTo = $junction->association($this->source()->alias());
+        $conds = $belongsTo->_joinCondition(['foreignKey' => $belongsTo->foreignKey()]);
+
+        $subquery = $this->find()
+            ->select(array_values($conds))
+            ->where($options['conditions']);
+
+        $subquery = $options['queryBuilder']($subquery);
+
+        $assoc = $this->junction()->association($this->target()->alias());
+        $conditions = $assoc->_joinCondition([
+            'foreignKey' => $this->targetForeignKey()
+        ]);
+        $subquery = $this->_appendJunctionJoin($subquery, $conditions);
+
+        $query
+            ->andWhere(function ($exp) use ($subquery, $conds) {
+                $identifiers = [];
+                foreach (array_keys($conds) as $field) {
+                    $identifiers = new IdentifierExpression($field);
```

Comment:
Does `IdentifierExpression` handle the field being a complex type, like a function expression?